Rtc mix

FRASTM opened the pull request:
stm32duino RTC add a parameter to set binary or bcd or Mix mode

FRASTM committed:
fix: harden prediv management vs clock config

prediv was not properly computed and some misalignment's
could occur depending of the RTC state at init.

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

FRASTM committed:
feat: support binary mode

Configure the alarm depending on the MIX mode
In case the RTC is running in MIX mode (binary and calendar),
the subsecond register is a 32-bit value (and not msec)

The Subsecond parameter is expressed in millisecond in
RTC_SetTime/GetTime RTC_StartAlarm/GetAlarm

Signed-off-by: Francois Ramu <francois.ramu@st.com>

fpistm commented:
Thanks

FRASTM commented:
Fixing compilation errors for other series than stm32wl, especially when the MIX mode is not defined.

FRASTM committed:
chore: provide a way to get the RTC handle

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm commented on src/rtc.c:
```diff
@@ -505,6 +505,9 @@ bool RTC_init(hourFormat_t format, rtcMode_t mode, sourceClock_t source, bool re
   HAL_RTCEx_EnableBypassShadow(&RtcHandle);
 #endif
 
+    HAL_NVIC_SetPriority(RTC_Alarm_IRQn, RTC_IRQ_PRIO, RTC_IRQ_SUBPRIO);
+    HAL_NVIC_EnableIRQ(RTC_Alarm_IRQn);
+
```
Don't think it is a good idea to add this. Arlarm IRQ should be enable only when alarm is set and enabled.
Which is done before by:
```
      if (isAlarmASet) {
        RTC_StartAlarm(ALARM_A, alarmDay, alarmHours, alarmMinutes, alarmSeconds, alarmSubseconds, alarmPeriod, alarmMask);
      }
#ifdef RTC_ALARM_B
      if (isAlarmBSet) {
        RTC_StartAlarm(ALARM_B, alarmBDay, alarmBHours, alarmBMinutes, alarmBSeconds, alarmBSubseconds, alarmBPeriod, alarmBMask);
      }
#endif
```

FRASTM commented:
export the RTC Handle to be used by the LoraWan
do not enable the Alarm IRQ
larmB is used by the LoraWan and set the MASK to RTC_ALARMMASK_ALL for calendar

FRASTM committed:
STM32RTC: syncAlarmTime Get the Alarm given by the parameter

In case the Alarm name is ALARM_B, this change will avoid
getting the Alarm_A

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
chore(examples): RTC running in MIX or BCD mode with Alarm

This examples is configuring the RTC in MIX (binary + calendar) or BCD
and sets Alarm A & B (if exists) few ms after the current time.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

fpistm committed:
fix: code syntax for better readability

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
feat: add a parameter to set binary mode

This is valid when the RTC_BINARY_MIX mode exists in the RTC
(bitfield in the RTC ICSR register)
Set the RTC mode through a setBinaryMode function to be called
before begin.

Signed-off-by: Francois Ramu <francois.ramu@st.com>
Co-authored-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
fix: init RtcHandle.instance

else this prevent isConfigured(void) to return correct value.

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
chore: set clock source and (a)synchronous prescalers together

(a)synchronous prescalers values depend on the RTC clock source.

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
ci(config): limit number of build

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
fix(example): enable alarm(s) if neeed

Formatted with Arduino astyle.

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

fpistm committed:
feat: add the HAL_MSP_Init/DeInit functions

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

FRASTM committed:
feat: RTC Time and Alarm configuration in BCD and BIN modes

define the Getime and startAlarm function when the RTC is configured
in BCD (MODE_BINARY_NONE) or BIN (MODE_BINARY_ONLY).
In MODE_BINARY_ONLY, Time and Date are not used, only subsecond

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
chore(examples): RTC running in BINary mode with Alarm

This examples is configuring the RTC in BIN mode (binary only)
and sets Alarm A & B (if exists) a few ms after the current time.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

fpistm committed:
fix: add missing else

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>

FRASTM committed:
feat: add new API to setAlarmTime with subsecond

Add new API to set alarm Time with subseconds which is a nb of
milliseconds. Similar to setAlarmSubSeconds.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

fpistm committed:
Merge pull request #98 from fpistm/prediv

harden prediv management and add parameter to set binary mode

fpistm committed:
chore(keywords): add new api and types

Signed-off-by: Frederic Pillon <frederic.pillon@st.com>
